### PR TITLE
Bug 2092047: 1.24 rebase

### PR DESCRIPTION
**Pull request by flavio-fernandes**
https://issues.redhat.com/browse/SDN-3038

### CODE REVIEW

**Comment by flavio-fernandes** [comment]
@ricky-rav PTAL

File: cmd/cloud-network-config-controller/main.go [hunk 1]
```diff
@@ -63,23 +67,44 @@ func main() {
 		klog.Exitf("Error building kubernetes clientset: %s", err.Error())
 	}
 
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(klog.Infof)
+	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: v1core.New(kubeClient.CoreV1().RESTClient()).Events("")})
+	component := "FIXME???"
```
Comment (flavio-fernandes) [inline cmd/cloud-network-config-controller/main.go:73]: need help on what value to put here!

File: cmd/cloud-network-config-controller/main.go [hunk 2]
```diff
@@ -63,23 +67,44 @@ func main() {
...
+	rl, err := resourcelock.New(
+		resourcelock.ConfigMapsLeasesResourceLock,
+		controllerNamespace,
+		resourceLockName,
+		kubeClient.CoreV1(),
+		kubeClient.CoordinationV1(),
+		resourcelock.ResourceLockConfig{
+			Identity:      controllerName,
+			EventRecorder: eventRecorder,
+		})
```
[the remark this answers is not in the document]
Comment (deads2k) [reply cmd/cloud-network-config-controller/main.go:85]: > @deads2k please help

Can you be more specific?  If leader election is failing, I think that the workloads team or apiserver team can help you.

[the remark this answers is not in the document]
Comment (flavio-fernandes) [reply cmd/cloud-network-config-controller/main.go:85]: ack. let me get more details and better explain myself. TY!

File: cmd/cloud-network-config-controller/main.go [hunk 3]
```diff
@@ -63,23 +68,33 @@ func main() {
...
+
+	rl, err := resourcelock.New(
+		resourcelock.ConfigMapsLeasesResourceLock,
+		controllerNamespace,
+		resourceLockName,
+		kubeClient.CoreV1(),
+		kubeClient.CoordinationV1(),
+		resourcelock.ResourceLockConfig{
+			Identity:      controllerName,
+			EventRecorder: eventRecorder,
```
Comment (andreaskaris) [inline cmd/cloud-network-config-controller/main.go:84]: EventRecorder is optional according to the API spec:
https://pkg.go.dev/k8s.io/client-go/tools/leaderelection/resourcelock#ResourceLockConfig

Any reason why you added this?

**Review by andreaskaris — APPROVED** [review]
/lgtm

According to my limited understanding. No idea about the switch from ConfigMapLock to ConfigMapsLeasesResourceLock and what the additional secondary leaselock does, but it seems to be required - perhaps good to ask in the meeting :-)
I didn't look at whatever go touched though, so let's hope that nobody hacked the vendor directory ;-)

================================================

new:
~~~
rl, err := resourcelock.New(
		resourcelock.ConfigMapsLeasesResourceLock,
		controllerNamespace,
		resourceLockName,
		kubeClient.CoreV1(),
		kubeClient.CoordinationV1(),
		resourcelock.ResourceLockConfig{
			Identity:      controllerName,
		})

if err != nil {
		klog.Exitf("Error building resource lock: %s", err.Error())
	}
~~~

vs old:
~~~
		Lock: &resourcelock.ConfigMapLock{
			ConfigMapMeta: metav1.ObjectMeta{
				Name:      resourceLockName,
				Namespace: controllerNamespace,
			},
			Client: kubeClient.CoreV1(),
			LockConfig: resourcelock.ResourceLockConfig{
				Identity: controllerName,
			},
		},
~~~

Looking at: https://github.com/kubernetes/client-go/blob/release-1.24/tools/leaderelection/resourcelock/interface.go#L166
~~~
func New(lockType string, ns string, name string, coreClient corev1.CoreV1Interface, coordinationClient coordinationv1.CoordinationV1Interface, rlc ResourceLockConfig) (Interface, error) {
	endpointsLock := &endpointsLock{
		EndpointsMeta: metav1.ObjectMeta{
			Namespace: ns,
			Name:      name,
		},
		Client:     coreClient,
		LockConfig: rlc,
	}
	configmapLock := &configMapLock{
		ConfigMapMeta: metav1.ObjectMeta{
			Namespace: ns,
			Name:      name,
		},
		Client:     coreClient,
		LockConfig: rlc,
	}
	leaseLock := &LeaseLock{
		LeaseMeta: metav1.ObjectMeta{
			Namespace: ns,
			Name:      name,
		},
		Client:     coordinationClient,
		LockConfig: rlc,
	}
	switch lockType {
	case endpointsResourceLock:
		return nil, fmt.Errorf("endpoints lock is removed, migrate to %s", EndpointsLeasesResourceLock)
	case configMapsResourceLock:
		return nil, fmt.Errorf("configmaps lock is removed, migrate to %s", ConfigMapsLeasesResourceLock)
	case LeasesResourceLock:
		return leaseLock, nil
	case EndpointsLeasesResourceLock:
		return &MultiLock{
			Primary:   endpointsLock,
			Secondary: leaseLock,
		}, nil
	case ConfigMapsLeasesResourceLock:
		return &MultiLock{
			Primary:   configmapLock,
			Secondary: leaseLock,
		}, nil
~~~

**Comment by flavio-fernandes** [comment]
/hold
We must not merge this before https://github.com/openshift/cluster-network-operator/pull/1461

File: cmd/cloud-network-config-controller/main.go (hunk 3 above)
Comment (flavio-fernandes) [reply cmd/cloud-network-config-controller/main.go]: No good reason. I removed it now. ;)

**Pull request merged**